### PR TITLE
typeck: fix `?` suggestion span

### PR DESCRIPTION
**Pull request by jieyouxu**
Noticed in <https://github.com/rust-lang/rust/pull/112043#issuecomment-2043565292>, if the

```
use the `?` operator to extract the `Result<(), std::fmt::Error>` value, propagating a `Result::Err` value to the caller
```

suggestion is applied to a macro that comes from a non-local crate (e.g. the stdlib), the suggestion span can become non-local, which will cause newer rustfix versions to fail.

This PR tries to remedy the problem by recursively probing ancestors of the expression span, trying to identify the most ancestor span that is (1) still local, and (2) still shares the same syntax context as the expression.

This is the same strategy used in https://github.com/rust-lang/rust/pull/112043.

The test unfortunately cannot `//@ run-rustfix` because there are two conflicting MaybeIncorrect suggestions that when collectively applied, cause the fixed source file to become non-compilable.

Also avoid running `//@ run-rustfix` for `tests/ui/typeck/issue-112007-leaked-writeln-macro-internals.rs` because that also contains conflicting suggestions.

cc @ehuss who noticed this. This question mark span fix + not running rustfix on the tests containing conflicting MaybeIncorrect suggestions should hopefully unblock rustfix from updating.

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs [hunk 1]
```diff
@@ -3180,3 +3171,19 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
     }
 }
+
+/// For suggestion span, recursively try to look for the ancestor span which shares the same syntax
+/// context as the initially provided `span`. This will recursively look into local macros until the
+/// span inside the most ancestor local macro is found. It will stop recursing as soon as the
+/// syntax context of a potential parent callsite changes, such as if the potential parent callsite
+/// is in a foreign macro. This helps to prevent leaking implementation details from upstream
+/// crates and stdlib crates that the user likely have no control over.
+fn find_local_most_ancestor_suggestable_span(initial_span: Span) -> Span {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:3181]: Why is this different than [`Span::find_ancestor_in_same_ctxt`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/span_encoding/struct.Span.html#method.find_ancestor_in_same_ctxt)?

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:3181]: Also https://github.com/rust-lang/rust/pull/112043/files#r1556509496

Comment (compiler-errors) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:3181]: Oh, I see. We find the *last* ancestor that shares the ctxt, not the first. Huh.

Comment (jieyouxu) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:3181]: Yeah, it was to get better error reporting for when there are multiple local nested macros.

Comment (cjgillot) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:3181]: This method should go in `rustc_span`, as a method on `Span`.

Comment (jieyouxu) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs:3181]: Do you have a better name for this? `find_local_most_ancestor_suggestable_span` is what I came up with but it feels not very accurate

**Comment by jieyouxu** [comment]
(Ignored the doctest because it contains type errors for illustration purposes)
@rustbot ready

File: compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs (hunk 1 above)
Comment (Nadrieril) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs]: How about `find_oldest_ancestor_in_same_ctxt`?

Comment (jieyouxu) [reply compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs]: Sounds better than what I had 😆

**Comment by jieyouxu** [comment]
Changed the method name to `find_oldest_ancestor_in_same_ctxt` and fixed its docs.

**Comment by Nadrieril** [comment]
Looks good!

@bors r+

**Pull request merged**